README.md

<p align="center">

<a href="https://github.com/italo-granato/BGGE">
<img src="Logo.png" alt="Bayesian Genomic Linear Models Applied to GE Genome Selection"/>
</a>

<h4 align="center">

Bayesian Genomic Linear Models Applied to GE Genome Selection -
Development version 0.6.5

</h4>

<h4 align="center">

\[Last README update: 2018-11-13\]

</h4>

<p align="center">

<a href="https://www.tidyverse.org/lifecycle/#maturing">
<img src="https://img.shields.io/badge/lifecycle-maturing-blue.svg" alt="Maturing">
</a> <a href="https://www.gnu.org/licenses/lgpl-3.0">
<img src="https://img.shields.io/badge/License-LGPL%20v3-blue.svg" alt="LGPL, Version 3.0">
</a> <a href="http://www.repostatus.org/#active">
<img src="http://www.repostatus.org/badges/latest/active.svg" alt="Status of the Repo: Active">
</a> <a href="">
<img src="http://cranlogs.r-pkg.org/badges/BGGE" alt="Dowloads from the CRAN">
</a> <a href="https://cran.r-project.org/package=BGGE">
<img src="http://www.r-pkg.org/badges/version-ago/BGGE" alt="CRAN"> </a>

</p>

</p>

# Table of contents

  - [NEWS](#news)
  - [Instructions](#instructions)
      - [Installation](#install)
      - [Load the package](#package)
      - [Quick use](#example1)
      - [Other Params](#params)
  - [How to cite this package](#cite)
  - [Contributions](#contributions)
  - [Authors](#authors)

<h2 id="news">

News of this version (0.6.5)

</h2>

Changes:

  - Added compatibility with the functions (print() and plot()).
  - Include ORCID of the authors.
  - Include url of the project.
  - Now the `BGGE()` function exports the input parameters.

See the last updates in [NEWS](NEWS.md).

<h2 id="instructions">

Instructions for proper implementation

</h2>

<h3 id="install">

Installation

</h3>

To complete installation of dev version of BGGE from GitHub, you must
have previously installed the devtools package.

``` r
install.packages('devtools')
devtools::install_github('italo/BGGE')
```

If you want to use the stable version of BGGE package, install it from
CRAN.

``` r
install.packages('BGGE')
```

<h3 id="package">

Load the package

</h3>

``` r
library(BGGE)
```

<h3 id="example1">

Example of simple usage of the package

</h3>

``` r
library(BGGE)
library(BGLR)
data(wheat)

X <- scale(wheat.X, scale = TRUE, center = TRUE)
rownames(X) <- 1:599
pheno_geno <- data.frame(env = gl(n = 4, k = 599), 
                         GID = gl(n=599, k=1, length = 599*4),
                         value = as.vector(wheat.Y))

# Creating kernel for GE model

K <- getK(Y = pheno_geno, X = X, kernel = "GB", model = "MM")
y <- as.vector(wheat.Y)

fit <- BGGE(y = y, K = K, ne = rep(599, 4))
fit
```

    ## Model Fitted with: 
    ##  1000  Iterations, burning the first  200  and thining every  3 
    ## 
    ##  Some predicted Values: 
    ##  [1]  -0.2737  -0.4081  -0.3799   0.3694  -0.0362   0.2354   0.2840
    ##  [8]  -0.1194  -0.7742  -1.0280
    ## 
    ##  Use str() function to found more datailed information.

``` r
plot(fit)
```

![](README_files/figure-gfm/BGGEUse-1.png)<!-- -->

<h3 id="params">

Others
params

</h3>

| params  | Use                                                                   |
| ------- | --------------------------------------------------------------------- |
| XF      | Design matrix for fixed effects.                                      |
| ite     | Number of iterations.                                                 |
| ne      | Number of subjects by environment.                                    |
| burn    | Number of iterations to be discarded as burn-in.                      |
| thin    | Thinin interval.                                                      |
| verbose | Should report be printed on screen?                                   |
| tol     | tolerance for zero. Default is 1e-10                                  |
| R2      | the proportion of variance expected to be explained by the regression |

<h2 id="cite">

Citation

</h2>

First option, by the paper.

``` tex
@article {Granato2018,
    author = {Granato, Italo and Cuevas, Jaime and Luna-V{\'a}zquez, Francisco J. and Crossa, Jos{\'e} and Montesinos-L{\'o}pez, Osval A. and Burgue{\~n}o, Juan and Fritsche-Neto, Roberto},
    title = {BGGE: A New Package for Genomic-Enabled Prediction Incorporating Genotype {\texttimes} Environment Interaction Models},
    year = {2018},
    doi = {10.1534/g3.118.200435},
    publisher = {G3: Genes, Genomes, Genetics},
    URL = {http://www.g3journal.org/content/early/2018/07/25/g3.118.200435},
    eprint = {http://www.g3journal.org/content/early/2018/07/25/g3.118.200435.full.pdf},
    journal = {G3: Genes, Genomes, Genetics}
}
```

Second option, by the package

``` r
citation('BGGE')
```

    ## 
    ## To cite package 'BGGE' in publications use:
    ## 
    ##   Italo Granato, Luna-Vázquez Francisco J. and Cuevas Jaime
    ##   (2018). BGGE: Bayesian Genomic Linear Models Applied to GE
    ##   Genome Selection. R package version 0.6.5.
    ##   https://github.com/italo-granato/BGGE
    ## 
    ## A BibTeX entry for LaTeX users is
    ## 
    ##   @Manual{,
    ##     title = {BGGE: Bayesian Genomic Linear Models Applied to GE Genome Selection},
    ##     author = {Italo Granato and Luna-Vázquez {Francisco J.} and Cuevas Jaime},
    ##     year = {2018},
    ##     note = {R package version 0.6.5},
    ##     url = {https://github.com/italo-granato/BGGE},
    ##   }

<h2 id="contributions">

Contributions

</h2>

If you have any suggestions or feedback, I would love to hear about it.
Feel free to report new issues in [this
link](https://github.com/italo/BGGE/issues/new), also if you want to
request a feature/report a bug, or make a pull request if you can
contribute.

<h2 id="contributions">

Research and implementations

</h2>

  - [BGGE: A New Package for Genomic-Enabled Prediction Incorporating
    Genotype × Environment Interaction
    Models](https://doi.org/10.1534/g3.118.200435)

<h2 id="authors">

Authors

</h2>

  - Italo Granato (Author, Maintainer)
  - Jaime D. Cuevas D. (Author)
  - Francisco J. Luna-Vázquez (Author, Maintainer)


NEWS.md
BGGE v0.6.5
==============

Changes:

* Added compatibility with the functions (print() and plot()).
* Include ORCID of the authors.
* Include url of the project.
* Now the `BGGE()` function exports the input parameters.


BGGE v0.6.4 (Release date: 2018-07)
==============

Changes:

* remove previously Sc conditional distribution
* fixed rm.na

BGGE v0.6.2 (Release date: 2018-05-10)
==============

Changes:

* Minor improvements.


BGGE v0.1
==============

Changes:

* Initial release

tests/testthat.R
library(testthat)
library(BGGE)

test_check("BGGE")


tests/testthat/test.R
library(testthat)
library(BGGE)
library(BGLR)

context('Kernel - With One Environment')
test_that('Generate Kernel with BGLR data', {
  data(wheat)
  
  X <- scale(wheat.X, scale = TRUE, center = TRUE)
  rownames(X) <- 1:599
  pheno_geno <- data.frame(env = 'One', 
                           GID = rownames(X),
                           value = wheat.Y[,1])
  
  # Expect error with multiple environments.
  expect_error(getK(Y = pheno_geno, X = X, kernel = "GB", model = "MM"))
  expect_error(getK(Y = pheno_geno, X = X, kernel = "GK", model = "MM"))
  
  # Test GB Kernel
  GB_Kernel <- getK(Y = pheno_geno, X = X, kernel = "GB", model = "SM")
  expect_output(str(GB_Kernel), 'List of 2')
  expect_equal(names(GB_Kernel), 'G')
  expect_is(GB_Kernel$G$Kernel, 'matrix')
  expect_equal(GB_Kernel$G$Type, 'D')
  
  # Test GK Kernel
  GK_Kernel <- getK(Y = pheno_geno, X = X, kernel = "GK", model = "SM")
  expect_output(str(GK_Kernel), 'List of 2')
  expect_equal(names(GK_Kernel), 'G')
  expect_is(GK_Kernel$G$Kernel, 'matrix')
  expect_equal(GK_Kernel$G$Type, 'D')
})


context('Kernel - Multi Environment')
test_that('Generate Kernel with BGLR data', {
  data(wheat)
  
  X <- scale(wheat.X, scale = TRUE, center = TRUE)
  rownames(X) <- 1:599
  pheno_geno <- data.frame(env = gl(n = 4, k = 599), 
                           GID = gl(n = 599, k = 1, length = 599 * 4),
                           value = as.vector(wheat.Y))
  
  # Expect error with single environments.
  expect_error(getK(Y = pheno_geno, X = X, kernel = "GB", model = "SM"))
  expect_error(getK(Y = pheno_geno, X = X, kernel = "GK", model = "SM"))
  
  # Test GB Kernel
  GB_Kernel <- getK(Y = pheno_geno, X = X, kernel = "GB", model = "MM")
  expect_output(str(GB_Kernel), 'List of 2')
  expect_equal(names(GB_Kernel), 'G')
  expect_is(GB_Kernel$G$Kernel, 'matrix')
  expect_equal(GB_Kernel$G$Type, 'D')
  
  # Test GK Kernel
  GK_Kernel <- getK(Y = pheno_geno, X = X, kernel = "GK", model = "MM")
  expect_output(str(GK_Kernel), 'List of 2')
  expect_equal(names(GK_Kernel), 'G')
  expect_is(GK_Kernel$G$Kernel, 'matrix')
  expect_equal(GK_Kernel$G$Type, 'D')
})

context('Fitting')

test_that('Fit and predict BGLR data', {
  data(wheat)
  
  X <- scale(wheat.X, scale = TRUE, center = TRUE)
  rownames(X) <- 1:599
  pheno_geno <- data.frame(env = gl(n = 4, k = 599), 
                           GID = gl(n = 599, k = 1, length = 599 * 4),
                           value = as.vector(wheat.Y))
  
  # Creating kernel for GE model
  K <- getK(Y = pheno_geno, X = X, kernel = "GB", model = "MM")
  
  fm <- BGGE(y = pheno_geno$value, K = K, ne = rep(599, 4))

  # Test fitted model
  expect_is(fm, 'BGGE')
  expect_output(str(fm), 'List of 9')
  expect_is(fm$y, 'numeric')
  expect_is(fm$yHat, 'numeric')
})


R/BGGE.R
#' Genotype x Environment models using regression kernel
#' 
#' BGGE function fits Bayesian regression for continuous observations through regression kernels
#'
#' @usage BGGE(y, K, XF = NULL, ne, ite = 1000, burn = 200, thin = 3, verbose = TRUE, 
#'             tol = 1e-10, R2 = 0.5)
#'
#' @param y Vector of data. Should be numeric and NAs are allowed.
#' @param K list A two-level list Specify the regression kernels (co-variance matrix). The former is the \code{Kernel},
#' where is included the regression kernels. The later is the \code{Type}, specifying if the matrix is either \code{D} Dense or
#' \code{BD} Block Diagonal. A number of regression kernels or random effects to be fitted are specified in this list.  
#' @param XF matrix Design matrix (\eqn{n \times p}) for fixed effects
#' @param ne vector Number of genotypes by environment. 
#' @param ite numeric Number of iterations.
#' @param burn numeric Number of iterations to be discarded as burn-in.
#' @param thin numeric Thinin interval.
#' @param verbose Should iteration history be printed on console? If TRUE or 1 then it is printed,
#' otherwise, if another number $n$ is choosen the history is printed every $n$ times. The default is \code{FALSE}  
#' @param tol a numeric tolerance level. Eigenvalues lower than \code{tol} are discarded. Default is \code{1e-10}.
#' @param R2 the proportion of variance expected to be explained by the regression.
#' 
#' @details
#' The goal is to fit genomic prediction models for continuous outcomes through Gibbs sampler. BGGE uses a proposal for dimension reduction
#' through an orthogonal transformation of observed data (y) as well as differential shrinkage because of the prior variance assigned 
#' to regression parameters. Further details on this approach can be found in Cuevas et al. (2014).
#' The primaty genetic model is
#' \deqn{y = g + e}
#' where \eqn{y} is the response, \eqn{g} is the unknown random effect and \eqn{e} is the residual effect.
#' You can specify a number of random effects \eqn{g}, as many as desired, through a list of regression kernels related to each random effect in the
#' argument \code{K}.
#' The structure of \code{K} is a two level list, where the first element on the second level is the Kernel and the second element is a definition of
#' type of matrix. There are two definitions, either matrix is \code{D} (dense) or \code{BD} (Block Diagonal). As we make the spectral decomposition 
#' on the kernels, for block diagonal matrices, we take advantage of its structure and make decomposition on the submatrices instead of one big matrix.
#' For example, the regression kernels should be an structure like K = list(list(Kernel = G, Type = "D"), list(Kernel = G, Type = "BD")). 
#' The definition of one matrix as a block diagonal must be followed by the number of subjects in each submatrix in the block diagonal,
#' present in the \code{ne}, which allows sub matrices to be drawn. Some genotype by environment models has the block diagonal matrix type or similar.
#' The genotype x environment deviation matrix in MDs model (Sousa et al., 2017) has the structure of block diagonal. 
#' Also, the matrices for environment-specific variance in MDe models (Sousa et al., 2017) if summed, can form a structure of block diagonal, 
#' where is possible to extract sub matrices for each environment. In the case of all kernel be of the dense type, \code{ne} is ignored. 
#' 
#' @return
#'  A list with estimated posterior means of residual and genetic variance component for each term in the linear model and the genetic value predicted. Also the 
#'  values along with the chains are released. 
#' 
#' 
#' @examples 
#' # multi-environment main genotypic model
#' library(BGLR)
#' data(wheat)
#' X<-wheat.X[1:200,1:600]  # Subset of 200 subjects and 600 markers
#' rownames(X) <- 1:200
#' Y<-wheat.Y[1:200,]
#' A<-wheat.A[1:200,1:200] # Pedigree
#' 
#' GB<-tcrossprod(X)/ncol(X)
#' K<-list(G = list(Kernel = GB, Type = "D"))
#' y<-Y[,1]
#' fit<-BGGE(y = y,K = K, ne = length(y), ite = 300, burn = 100, thin = 2)
#' 
#' # multi-environment main genotypic model
#' Env <- as.factor(c(2,3)) #subset of 2 environments
#' pheno_geno <- data.frame(env = gl(n = 2, k = nrow(Y), labels = Env),
#'                          GID = gl(n = nrow(Y), k = 1,length = nrow(Y) * length(Env)),
#'                          value = as.vector(Y[,2:3]))
#' 
#' K <- getK(Y = pheno_geno, X = X, kernel = "GB", model = "MM")
#' y <- pheno_geno[,3]
#' fit <- BGGE(y = y, K = K, ne = rep(nrow(Y), length(Env)), ite = 300, burn = 100,thin = 1)
#' 
#' 
#' @seealso 
#' \code{\link[BGLR]{BGLR}}
#' 
#' @references
#' Cuevas, J., Perez-Elizalde, S., Soberanis, V., Perez-Rodriguez, P., Gianola, D., & Crossa, J. (2014).
#' Bayesian genomic-enabled prediction as an inverse problem. G3: Genes, Genomes, Genetics, 4(10), 1991-2001.
#' 
#' Sousa, M. B., Cuevas, J., Oliveira, E. G. C., Perez-Rodriguez, P., Jarquin, D., Fritsche-Neto, R., Burgueno, J. 
#' & Crossa, J. (2017). Genomic-enabled prediction in maize using kernel models with genotype x environment interaction.
#' G3: Genes, Genomes, Genetics, 7(6), 1995-2014.
#' 
#' 
#' @export
BGGE <- function(y, K, XF = NULL, ne, ite = 1000, burn = 200, thin = 3, verbose = FALSE, tol = 1e-10, R2 = 0.5) {
  
  ### PART I  - Conditional distributions functions and eigen descomposition ####
  # Conditional Distribution of tranformed genetic effects b (U'u)
  #' @import stats
  dcondb <- function(n, media, vari) {
    sd <- sqrt(vari)
    return(rnorm(n, media, sd))
  }
  
  # Conditional distribution of compound variance of genetic effects (sigb)
  dcondsigb <- function(b, deltav, n, nu, Sc) {
    z <- sum(b ^ 2 * deltav)
    return(1 / rgamma(1, (n + nu) / 2, (z + nu * Sc) / 2))
  } 
  
  # Conditional distribution of residual compound variance 
  dcondsigsq <- function(Aux, n, nu, Sce) {
    return(1 / rgamma(1, (n + nu) / 2, crossprod(Aux) / 2 + Sce / 2))
  }
  
  # Conditional fixed effects
  rmvnor <- function(n,media,sigma){
    z <- rnorm(n)
    return( media + crossprod(chol(sigma), z))
  }
  
  # Function for eigen descompositions
  eig <- function(K, tol) {
    ei <- eigen(K)
    fil <- which(ei$values > tol)
    return(list(ei$values[fil], ei$vectors[, fil]))
  }
  
  # Set spectral decomposition
  setDEC <- function(K, tol, ne) {
    
    sK <- vector("list", length = length(K))
    typeM <- sapply(K, function(x) x$Type)
    
    if (!all(typeM %in% c("BD", "D")))
      stop("Matrix should be of types BD or D")
    
    if (missing(ne)) {
      if (any(typeM == "BD"))
        stop("For type BD, number of subjects in each sub matrices should be provided")
    } else {
      if (length(ne) <= 1 & any(typeM == "BD"))
        stop("ne invalid. For type BD, number of subjects in each sub matrices should be provided")
      
      nsubK <- length(ne)
      
      if (nsubK > 1) {
        posf <- cumsum(ne)
        posi <- cumsum(c(1,ne[-length(ne)]))
      }
    }
    
    
    for (i in seq_len(K)) {
      if (K[[i]]$Type == "D") {
        tmp <- list()
        ei <- eig(K[[i]]$Kernel, tol)
        tmp$s <- ei[[1]]
        tmp$U <- ei[[2]]
        tmp$tU <- t(ei[[2]])
        tmp$nr <- length(ei[[1]])
        tmp$type <- "D"
        tmp$pos <- NA
        sK[[i]] <- list(tmp)
      }
      
      if (K[[i]]$Type == "BD") {
        cont <- 0
        tmp <- list()
        for (j in 1:nsubK) {
          Ktemp <- K[[i]]$Kernel[(posi[j]:posf[j]), (posi[j]:posf[j])]
          ei <- eig(Ktemp, tol)
          if (length(ei[[1]]) != 0) {
            cont <- cont + 1
            tmp[[cont]] <- list()
            tmp[[cont]]$s <- ei[[1]]
            tmp[[cont]]$U <- ei[[2]]
            tmp[[cont]]$tU <- t(ei[[2]])
            tmp[[cont]]$nr <- length(ei[[1]])
            tmp[[cont]]$type <- "BD"
            tmp[[cont]]$pos <- c(posi[j], posf[j])
          }
        }
        
        if (length(tmp) > 1) {
          sK[[i]] <- tmp
        } else {
          sK[[i]] <- list(tmp[[1]])
        }
      }
    }
    return(sK)
  }
  
  # verbose part I
  if (as.numeric(verbose) != 0) {
    cat("Setting parameters...", "\n", "\n")
  }
  
  ### PART II  Preparation for Gibbs sampler ######
  # Identification of NA's values
  y <- as.numeric(y)
  yNA <- is.na(y)
  whichNa <- which(yNA)
  nNa <- length(whichNa)
  mu <- mean(y, na.rm = TRUE)
  n <- length(y)
  
  if (nNa > 0) { 
    y[whichNa] <- mu
  }
  
  # name of each kernel (important to following procedures)
  if (is.null(names(K))) {
    names(K) <- paste("K", seq(length(K)), sep = "")
  }
  
  # initial values of fixed effects 
  if (!is.null(XF)) {
    Bet <- solve(crossprod(XF), crossprod(XF, y))
    nBet <- length(Bet)
    tXX <- solve(crossprod(XF))
  }
  
  # Eigen descomposition for nk Kernels
  nk <- length(K)
  nr <- numeric(length(K))
  typeM <- sapply(K, function(x) x$Type)
  
  if (!all(typeM %in% c("BD", "D")))
    stop("Matrix should be of types BD or D")
  
  if (length(ne) == 1 & any(typeM == "BD"))
    stop("Type BD should be used only for block diagonal matrices")
  
  Ei <- setDEC(K = K, ne = ne, tol = tol)
  
  # Initial values for Monte Carlo Markov Chains (MCMC)
  
  nCum <- sum(seq(1, ite) %% thin == 0)
  
  chain <- vector("list", length = 3)
  names(chain) <- c("mu", "varE", "K")
  chain$varE <- numeric(nCum)
  chain$mu <- numeric(nCum)
  
  chain$K <- vector("list", length = nk)
  names(chain$K) <- names(K)
  chain$K[seq(nk)] <- list(list(varU = numeric(nCum)))
  
  cpred <- vector('list', length = nk)
  names(cpred) <- names(K)
  cpred[seq(nk)] <- list(U = matrix(NA_integer_, nrow = nCum, ncol = n))
  
  
  nu <- 3
  Sce <- (nu + 2) * (1 - R2) * var(y, na.rm = TRUE)
  Sc <- numeric(length(K))
  for (i in seq_len(K)) {
    Sc[i] <- (nu + 2) * R2 * var(y, na.rm = TRUE)/mean(diag(K[[i]]$Kernel))
  }
  tau <- 0.01
  u <- list()
  for (j in 1:nk) {
    u[[j]] <- rnorm(n, 0, 1 / (2 * n))
  }
  sigsq <- var(y)
  #Sc <- rep(0.01, nk)
  sigb <- rep(0.2, nk)
  
  temp <- y - mu
  
  if (!is.null(XF)) {
    B.mcmc <- matrix(0, nrow = nCum, ncol = nBet)
    temp <- temp - XF %*% Bet
  }

  temp <- temp - Reduce('+', u)
  nSel <- 0
  i <- 1
  
  ### PART III  Fitted model with training data ####
  
  # Iterations of Gibbs sampler
  while (i <= ite) {
    time.init <- proc.time()[3]
    
    # Conditional of mu
    temp <- temp + mu
    mu <- rnorm(1, mean(temp), sqrt(sigsq/n))
    #mu.mcmc[i] <- mu
    temp <- temp - mu
    
    # Conditional of fixed effects
    if (!is.null(XF)) {
      temp <- temp + XF %*% Bet
      vari <- sigsq * tXX
      media <- tXX %*% crossprod(XF, temp)
      Bet <- rmvnor(nBet, media, vari)
      temp <- temp - XF %*% Bet
    }
    
    # Conditionals x Kernel
    for (j in 1:nk) {
      # Sampling genetic effects
      
      if (typeM[j] == "D") {
        temp <- temp + u[[j]]
        d <- crossprod(Ei[[j]][[1]]$U, temp)
        s <- Ei[[j]][[1]]$s
        deltav <- 1 / s
        lambda <- sigb[j]
        vari <- s * lambda / (1 + s * lambda * tau)
        media <- tau * vari * d
        nr <- Ei[[j]][[1]]$nr
        b <- dcondb(nr, media, vari)
        u[[j]] <- crossprod(Ei[[j]][[1]]$tU ,b)
        temp <- temp - u[[j]]
      }
      
      if (typeM[j] == "BD") {
        nsk <- length(Ei[[j]])
        
        if (length(nsk > 1)) {
          temp <- temp + u[[j]]
          d <- NULL
          s <- NULL
          neiv <- numeric(nsk)
          pos <- matrix(NA, ncol = 2, nrow = nsk)
          for (k in 1:nsk) {
            pos[k,] <- Ei[[j]][[k]]$pos
            d <- c(d, crossprod(Ei[[j]][[k]]$U, temp[pos[k, 1]:pos[k, 2]]))
            neiv[k] <- length(Ei[[j]][[k]]$s)
            s <- c(s, Ei[[j]][[k]]$s)
          }
          deltav <- 1/s
          lambda <- sigb[j]
          vari <- s * lambda / (1 + s * lambda * tau)
          media <- tau*vari*d
          nr <- length(s)
          b <- dcondb(nr, media, vari)
          
          posf <- cumsum(neiv)
          posi <- cumsum(c(1, neiv[-length(neiv)]))
          utmp <- numeric(n)
          for (k in 1:nsk) {
            utmp[pos[k, 1]:pos[k, 2] ] <- crossprod(Ei[[j]][[k]]$tU, b[posi[k]:posf[k] ])
          }
          u[[j]] <- utmp
          temp <- temp - u[[j]]
          
        }else{
          temp <- temp + u[[j]]
          pos <- Ei[[j]]$pos
          d <- crossprod(Ei[[j]][[1]]$U, temp[pos[1]:pos[2]])
          s <- Ei[[j]][[1]]$s
          deltav <- 1/s
          lambda <- sigb[j]
          vari <- s * lambda / (1 + s * lambda * tau)
          media <- tau*vari*d
          nr <- Ei[[j]][[1]]$nr
          b <- dcondb(nr, media, vari)
          utmp <- numeric(n)
          utmp[pos[1]:pos[2]] <-  crossprod(Ei[[j]][[1]]$tU, b)
          u[[j]] <- utmp
          temp <- temp - u[[j]]
        }
      }
      
      # Sampling scale hyperparameters and variance of genetic effects
      sigb[j] <- dcondsigb(b, deltav, nr, nu, Sc[j])
    }
    
    # Sampling residual variance 
    res <- temp
    #Sce <- dcondSc(nu, sigsq)
    sigsq <- dcondsigsq(res, n, nu, Sce)
    tau <- 1 / sigsq
    
    # Predicting missing values
    if (nNa > 0) {
      uhat <- Reduce('+', u)
      
      if (!is.null(XF)) {
        aux <- XF[yNA,] %*% Bet
      }else{
        aux <- 0
      }
      
      y[whichNa] <- aux + mu + uhat[whichNa] + rnorm(n = nNa, sd = sqrt(sigsq))
      temp[whichNa] <- y[whichNa] - uhat[whichNa] - aux - mu
    }
    
    # Separating what is for the chain
    if (i %% thin == 0) {
      nSel <- nSel + 1
      chain$varE[nSel] <- sigsq
      chain$mu[nSel] <- mu
      if (!is.null(XF)) {
        B.mcmc[nSel,] <- Bet
      }
      for (j in 1:nk) {
        cpred[[j]][nSel,] <- u[[j]]
        chain$K[[j]]$varU[nSel] <- sigb[j]
      }
    }
    
    # Verbose 
    if (as.numeric(verbose) != 0 & i %% as.numeric(verbose) == 0) {
      time.end <- proc.time()[3]
      cat("Iter: ", i, "time: ", round(time.end - time.init, 3),"\n")
    }
    i <- i + 1
  }

  
  ###### PART IV  Output ######
  #Sampling
  draw <- seq(ite)[seq(ite) %% thin == 0] > burn
  
  mu.est <- mean(chain$mu[draw])
  yHat <- mu.est
  
  if (!is.null(XF)) {
    B <- colMeans(B.mcmc[draw,])
    yHat <- yHat + XF %*% B
  }
  
  u.est <- sapply(cpred, FUN = function(x) colMeans(x[draw, ]) )
  yHat <- yHat + rowSums(u.est)
  
  out <- list()
  out$yHat <- yHat
  out$varE <- mean(chain$varE[draw])
  out$varE.sd <- sd(chain$varE[draw])
  
  out$K <- vector('list', length = nk)
  names(out$K) <- names(cpred)
  for (i in 1:nk) {
    out$K[[i]]$u <- colMeans(cpred[[i]][draw, ])
    out$K[[i]]$u.sd <- apply(cpred[[i]][draw, ], MARGIN = 2, sd)
    out$K[[i]]$varu <- mean(chain$K[[i]]$varU[draw])
    out$K[[i]]$varu.sd <- sd(chain$K[[i]]$varU[draw])
  }
  
  out$chain <- chain
  out$ite <- ite
  out$burn <- burn
  out$thin <- thin
  out$model <- K$model
  out$kernel <- K$kernel
  out$y <- y
  class(out) <- "BGGE"
  return(out)
}


R/mainGE2.R
#' Genotype x Environment models using linear or gaussian kernel
#'
#' @usage mainGE(Y, X, XF=NULL, W=NULL, kernel=c("GK", "GB"), K = NULL, 
#'              h=1, model = c("SM", "MM", "MDs", "MDe", "Cov"), ite = 1000, burn = 300,
#'              thin = 5, verbose = FALSE, tol = 1e-10, ...)
#'
#' @param Y \code{data.frame} Phenotypic data with three columns. The first column assigns the environment,
#' the second column assigns individuals and the third column contains the trait of interest.
#' @param X \code{matrix} Marker matrix with individuals in rows and marker in columns
#' @param XF \code{matrix} Design matrix (\eqn{n \times p}) for fixed effects
#' @param W \code{matrix} Environmental covariance matrix of dimension (\eqn{n \times n}). If \code{W} is provided, it can be used along with \code{MM} and \code{MDs} models. See details
#' @param kernel  regression kernel to be used. Methods currently implemented are the gaussian kernel \code{GK} and the linear kernel \code{G-BLUP}
#' @param K \code{matrix} Single kernel matrix in case it needs to use a different kernel from those supported.
#' @param h \code{numeric} Bandwidth parameter to create the gaussian kernel. For \code{method} \code{GK}, if \code{h} is not provided,
#' then it is computed following a empirical bayesian selection method. See details
#' @param model Specifies the genotype by environment model to be fitted. See details
#' @param ite \code{integer} Number of iterations.
#' @param burn \code{integer} Number of iterations to be discarded as burn-in.
#' @param thin \code{integer} Thinin interval.
#' @param verbose \code{logical} Should report be printed on screen?
#' @param tol \code{numeric} tolerance for zero. Default is 1e-10
#' @param \dots additional arguments to be passed.

#' @details
#' The goal is to fit genomic prediction models including GxE interaction. For this, is necessary kernel matrices for diferents models currently supported.
#' These models can be adjusted using different kernels. \code{GB} creates a linear kernel resulted from the cross-product of centered and standarized marker genotypes divide by the number of markers \eqn{p}:
#'  \deqn{G = \frac{XX^T}{p}}
#' Other kernel option supported is the Gaussian Kernel \code{GK}, estimated as:
#'  \deqn{ K (x_i, x_{i'}) = exp(-h d_{ii'}^2)}
#' However other kernels can be provided through \code{K}. In this case, arguments \code{X}, \code{kernel} and \code{h} are ignored. See \code{\link{h.fun}}
#' 
#' mainGE uses the BLMMD function and the package MTM to fit the current models:
#' \itemize{
#' \item \code{SM}: is the single-environment main genotypic effect model - The SM model fits the data for each single environment separately.
#' \item \code{MM}: is the multi-environment main genotypic effect model - Multi-environment model considering the main random genetic effects across environments.
#' \item \code{MDs}: is the multi-environment single variance genotype × environment deviation model - This model is an extension of \code{MM} by adding the random interaction effect of the environments
#' with the genetic information of genotypes.
#' \item \code{MDe}: is the multi-environment, environment-specific variance genotype × environment deviation model - This model separates the genetic effects of markers into the main marker effects (across environments) and the specific marker effects (for each environment).
#' \item \code{Cov}: is the multi-environment, variance-covariance environment-specific model -
#' }
#'
#' @return
#' A list with estimated posterior means of variance components for each term in the linear model and the genetic value predicted.
#' 
#'
#' @seealso \code{\link[MTM]{MTM}}, \code{\link{getK}} and \code{\link{BLMMD}}
#'
#'@examples
#' # create kernel matrix for model MDs using wheat dataset
#' library(BGLR)
#' 
#' data(wheat)
#' X <- scale(wheat.X, scale = TRUE, center = TRUE)
#' rownames(X) <- 1:599
#' pheno_geno <- data.frame(env = gl(n = 4, k = 599), 
#'                GID = gl(n=599, k=1, length = 599*4),
#'                value = as.vector(wheat.Y))
#' 
#' fit <- mainGE(Y = pheno_geno, X = X, kernel = "GB", model = "MDs")
#'  
#'
#'@export
mainGE <- function(Y, X, XF=NULL, W=NULL, kernel=c("GK", "GB"), K = NULL, h=1, model = c("SM", "MM", "MDs", "MDe", "Cov"),
                   ite = 1000, burn = 300, thin = 5, verbose = FALSE, tol = 1e-10, ...) {
  
  hasW <- !is.null(W)
  
  names(Y) <- c("env", "subjects", "value")
  subj <- levels(Y$subjects)
  env <- levels(Y$env)
  
  #setting kernels
  setK <- getK(Y = Y, X = X, K = K, kernel = kernel, h = h, model = model)
  y <- as.vector(Y[,3L])
  
  if (model %in% c("MM", "MDs") & hasW) {
    Ze <- model.matrix(~factor(Y[,1])-1)
    Zg <- model.matrix(~factor(Y[,2])-1)
    EC <- Ze %*% tcrossprod(W, Ze)
    GEC <- setK$G * EC
    setK <- c(setK, list(EC = list(K = EC), GEC = list(K = GEC)))
  }
  
  if (model == "Cov") {
    #' @importFrom reshape2 acast
    tmpY <- acast(Y, subjects ~ env, add.missing=TRUE, value = "value")
    
    fit <- GEcov(Y = tmpY, K = setK$G, nIter = ite, burnIn = burn, thin = thin,...)
  }
  else {
     fit <- BLMMD(y = y, K = setK, XF = XF, ite = ite, burn = burn, thin = thin, verbose = verbose, tol = tol)
  }
  return(fit)
}


GEcov <- function(Y, K, nIter = 1000, burnIn = 300, thin = 5, ...)
{
  env <- ncol(Y)
  In <- diag(x = 1, nrow = nrow(Y), ncol = nrow(Y))
  
  # How to find df0 and S0?
  ETA <- list(list( K = K, COV = list(type = 'UN', df0 =  env, S0 = diag(env))),
              list(K = In, COV = list(type = 'UN', df0 = env, S0 = diag(env))))
  resCov <- list( type = 'DIAG', S0 = rep(1, env), df0 = rep(1, env))
  
  #' @importFrom MTM MTM
  fit <- MTM(Y = Y, K = ETA, resCov = resCov, nIter = nIter, burnIn = burnIn, thin = thin,...)
  
  return(fit)
}
